fri/src/matrix_reducer.rs

```diff
@@ -0,0 +1,188 @@
+use alloc::vec;
+use alloc::vec::Vec;
+use itertools::Itertools;
+use p3_field::{AbstractField, ExtensionField, Field, PackedField};
+use p3_matrix::MatrixRows;
+use p3_maybe_rayon::{IndexedParallelIterator, MaybeIntoParIter, ParallelIterator};
+use p3_util::indices_arr;
+use tracing::{info_span, instrument};
+
+// seems to be the sweet spot, could be tweaked based on benches
+const BATCH_SIZE: usize = 8;
+
+/// Optimized matrix reducer. Only works for binomial extension fields, or extension fields
+/// in which multiplication and addition with a base field element are simple elementwise operations.
+pub(crate) struct MatrixReducer<F: Field, EF> {
+    alpha: EF,
+    alpha_pow_width: EF,
+    transposed_alphas: Vec<[F::Packing; BATCH_SIZE]>,
+    current_alpha_pow: EF,
+}
+
+impl<F: Field, EF: ExtensionField<F>> MatrixReducer<F, EF> {
+    pub(crate) fn new(alpha: EF) -> Self {
+        let alpha_pows = alpha
+            .powers()
+            .take(F::Packing::WIDTH * BATCH_SIZE)
+            .collect_vec();
+        let transposed_alphas = (0..EF::D)
+            .map(|i| {
+                indices_arr::<BATCH_SIZE>().map(|j| {
+                    F::Packing::from_fn(|k| {
+                        alpha_pows[j * F::Packing::WIDTH + k].as_base_slice()[i]
+                    })
+                })
+            })
+            .collect_vec();
+        Self {
+            alpha,
+            alpha_pow_width: alpha.exp_u64(F::Packing::WIDTH as u64),
+            transposed_alphas,
+            current_alpha_pow: EF::one(),
+        }
+    }
+
+    #[instrument(name = "fold in matrices", level = "debug", skip(self, reduced, mats))]
+    pub(crate) fn reduce_matrices<M: MatrixRows<F> + Sync>(
+        &mut self,
+        reduced: &mut [EF],
+        height: usize,
+        mats: &[M],
+    ) {
+        // precompute alpha_pows, since we are not doing horner
```

Comment:
Horner on these made it really annoying for fri verifier to deal with multiple degree bounds